src/System.Net.Sockets/src/System/Net/Sockets/SocketPal.Windows.cs

```diff
@@ -708,61 +744,97 @@ public static SocketError Poll(SafeCloseSocket handle, int microseconds, SelectM
+                // 
+                // Negative microsecond values that weren't exactly (-1) were originally successfully 
+                // converted to a timeval struct containing unsigned non-zero integers.  This code 
+                // retains that behavior so that any app working around the original bug with, 
+                // for example, (-2) specified for microseconds, will continue to get the same behavior.
+
+                int socketCount;
+                fixed (IntPtr* readPtr = &MemoryMarshal.GetReference(readfileDescriptorSet))
+                fixed (IntPtr* writePtr = &MemoryMarshal.GetReference(writefileDescriptorSet))
+                fixed (IntPtr* errPtr = &MemoryMarshal.GetReference(errfileDescriptorSet))
+                {
+                    if (microseconds != -1)
+                    {
+                        Interop.Winsock.TimeValue IOwait = new Interop.Winsock.TimeValue();
+                        MicrosecondsToTimeValue((long)(uint)microseconds, ref IOwait);
+
+                        socketCount =
+                            Interop.Winsock.select(
+                                0, // ignored value
+                                readPtr,
+                                writePtr,
+                                errPtr,
+                                ref IOwait);
+                    }
+                    else
+                    {
+                        socketCount =
+                            Interop.Winsock.select(
+                                0, // ignored value
+                                readPtr,
+                                writePtr,
+                                errPtr,
+                                IntPtr.Zero);
+                    }
+                }
+                if (NetEventSource.IsEnabled)
+                    NetEventSource.Info(null, $"Interop.Winsock.select returns socketCount:{socketCount}");
 
-            if (NetEventSource.IsEnabled) NetEventSource.Info(null, $"Interop.Winsock.select returns socketCount:{socketCount}");
+                if ((SocketError)socketCount == SocketError.SocketError)
+                {
+                    return GetLastSocketError();
+                }
 
-            if ((SocketError)socketCount == SocketError.SocketError)
+                Socket.SelectFileDescriptor(checkRead, readfileDescriptorSet);
+                Socket.SelectFileDescriptor(checkWrite, writefileDescriptorSet);
+                Socket.SelectFileDescriptor(checkError, errfileDescriptorSet);
+
+                return SocketError.Success;
+            }
+            finally
             {
-                return GetLastSocketError();
+                if (leaseRead != null) ArrayPool<IntPtr>.Shared.Return(leaseRead);
+                if (leaseWrite != null) ArrayPool<IntPtr>.Shared.Return(leaseWrite);
+                if (leaseError != null) ArrayPool<IntPtr>.Shared.Return(leaseError);
```

Comment:
Do we have any tests currently that validate the > threshold code paths?  If not, mind adding one/some?

Reply:
will check and add if needed

Reply:
Thanks.  It looks like we do, actually:
https://github.com/dotnet/corefx/blob/f6c6a894bd5b385636d4db778cd3ce6bf09daca3/src/System.Net.Sockets/tests/FunctionalTests/SelectTest.cs#L27-L32

Reply:
dammit, I missed those - have just committed more! includes a more involved multi-segment read/write test, too; but I can perhaps kill the Select ones? or leave them?

Reply:
> dammit, I missed those - have just committed more! includes a more involved multi-segment read/write test, too; but I can perhaps kill the Select ones? or leave them?

Why don't you put up a new PR with the additional tests, and we can decide then if it should also delete / merge with the existing ones?

Reply:
meh, 'tis merged; whatever :)